### PR TITLE
Remove the base URI for docker.io images

### LINKED ISSUE #37912: allow usage of mirror for docker.io images
### Description

When I use devservices in a CI system, I have sometime a error due to the [docker hub rate limit](https://www.docker.com/increase-rate-limits/).

### Implementation ideas

The  images configured by default in quarkus use a full path.
So the `--registry-mirror=` option does not work.

As `docker.io` is the default registry for Docker and other runtimes like Podman, we can remove the prefix.

### PR DESCRIPTION
**Pull request by ggrebert**
This feature allow users to use a mirror registry
to bypass the docker hub rate limit.

fix #37912

### CODE REVIEW

File: extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/KafkaDevServicesBuildTimeConfig.java [hunk 1]
```diff
@@ -47,7 +47,7 @@ public class KafkaDevServicesBuildTimeConfig {
     public Provider provider = Provider.REDPANDA;
 
     public enum Provider {
-        REDPANDA("docker.io/vectorized/redpanda:v22.3.4"),
+        REDPANDA("vectorized/redpanda:v22.3.4"),
         STRIMZI("quay.io/strimzi-test-container/test-container:latest-kafka-3.2.1"),
         KAFKA_NATIVE("quay.io/ogunalp/kafka-native:latest");
```
Comment (gastaldi) [inline extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/KafkaDevServicesBuildTimeConfig.java:52]: Shouldn't the same apply to other registries as well?

**Comment by gsmet** [comment]
@cescoffier @holly-cummins @sanne could you have a look at this one? I remember we added the `docker.io` registry specifically at some point. Don't remember exactly why but I wonder if it was related to Podman.

**Comment by holly-cummins** [comment]
I'd forgotten all about this, but https://github.com/quarkusio/quarkus/commit/57332c3e53dd966de444f51970aa995806bc5d64 confirms the decision to be explicit about the name *was* related to Podman, and short names. It was an issue that @Sanne found.

File: extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/KafkaDevServicesBuildTimeConfig.java (hunk 1 above)
Comment (ggrebert) [reply extensions/kafka-client/deployment/src/main/java/io/quarkus/kafka/client/deployment/KafkaDevServicesBuildTimeConfig.java:52]: For `quay.io` it is more complicated: the `--registry-mirror` option works only with relative image.

But we can had a property to replace this value. But in an other PR,

**Comment by ggrebert** [comment]
> Yes, the registry name was required by podman. I'm finding references about something also required on fedora 35+.
> 
> Also, aren't all these images customizable by the user? Using the prefix also guarantees the origin of the image (and somewhat avoids having someone publishing a tainted image on a mirror).

Image are customizable by the user (actual workaround).
But (to my known and searches) there is not a prefix property to change the DNS of devservices images.

**Comment by cescoffier** [comment]
I'm not sure a prefix would help:

- should it be added to all images without prefix?
- should all images be prefix-less, but what do we do for images from quay or google?

**Comment by ggrebert** [comment]
> * should it be added to all images without prefix?

For me no. Because images without prefix are hosted on `docker.io` and can be mirrored directly by docker or podman.

> * should all images be prefix-less, but what do we do for images from quay or google?

Images hosted on `quay.io` and `gcr.io` have no rate limit. And some are not replicated on `docker.io`, so we cannot use prefix-less for those.
But it could be interesting to have a property to change easily the DNS (for internet less networks).

```properties
quarkus.devservices.registry.quay.dns=quay.io
quarkus.devservices.registry.gcr.dns=gcr.io
```

**Comment by cescoffier** [comment]
To do that, we need to parse the image names and replace the DNS (by the way, I'm not sure of the wording here; I would prefer something like `registry-hostname` (and even hostname is not perfect as you can pass IPs).

The next question is about the scope. Should it be global (like in your proposal) or per image (which is literally what we have today, as you can replace the image)? I like the global approach as it's more in phase with the idea of a mirror. However, as I'm not an expert on image registry mirroring, don't we have the risk of having a mirror containing some images but not others (like if some images are private, the mirror may not have them)? 

Finally, when the user overrides the image name, should we still apply the hostname? I would say no, but let's make sure we all agree.

**Comment by ggrebert** [comment]
new proposal:

```properties
quarkus.devservices.registry.quay.host=localhost:5000
quarkus.devservices.registry.gcr.host=gcr.mirror.internal
```

For private images, we already have to override the image property. So I am in tune with you:

```gherkin
When we override the image name
And the value use a full path
Then pull the image form this location
```

```gherkin
When we override the image name
And the value not use a full path
Then use the docker/podman configuration to pull the image
   (`docker.io` if no mirror is set)
```

So:
- if a user wants to pull a custom image from the quay.io registry, he must use a full path.
- if a user wants to pull a custom image from the docker.io registry, he can use full path ou relative path.

The good: this solution don't change the current behaviour)

**Comment by cescoffier** [comment]
Completely agree with this new proposal.

**Comment by ggrebert** [comment]
cool,

What do you prefer ? :

1. I rebase this PR (and check if no new image have been haded) and merge it to implement the docker.io/mirror part and I do a new one for the other registries ?
2. edit this PR which implement all the features

**Comment by ggrebert** [comment]
to me, I prefer the fisrt one

**Comment by cescoffier** [comment]
I would have done everything in this PR, but we can start with docker.io only and then do another PR with the other registries (less critical as it does not have rate limits)

**Comment by ggrebert** [comment]
ready to merge for me.

Issue created for the other registries: https://github.com/quarkusio/quarkus/issues/38146

**Comment by cescoffier** [comment]
Why didn't you implement what we discussed:

- a new property to change the host name of the docker.io registry
- keep the docker.io name and apply the new host

**Comment by cescoffier** [comment]
@ggrebert ping?

**Comment by cescoffier** [comment]
Yes, I think it's better, as it makes things more homogenous with the other registries.

**Comment by ggrebert** [comment]
OK, thx @cescoffier for clarifying the subject.

It can be take time to me due to all the different way the docker images are set in all the extensions and i work on it only on my free time.
But I am on it :metal:

**Comment by gsmet** [comment]
@ggrebert any progress on this one?

**Comment by ggrebert** [comment]
Hello @cescoffier and @gsmet,

Sorry, I really haven't had time to work on this as much as I would have hoped
(a complicated divorce that takes up all my free time...).

However, I still studied @cescoffier solution.
But it is not so easy to implement:

I was going to create a helper in [ConfigureUtil.java](https://github.com/quarkusio/quarkus/blob/main/extensions/devservices/common/src/main/java/io/quarkus/devservices/common/ConfigureUtil.java) which replaces the host.

But we then have to modify all the DevServices in each extension.
The real problem with this solution is for external extensions such as Quarkiverse.
And has you know, the Quarkiverse extentions don't use the latest Quarkus version,
So all extensions must update Quarkus before implementing the solution.

Note:
Tthere already exist docker.io images without host (solution based on registry-mirror) as proposed in the original PM.
[example for MySQL](https://github.com/quarkusio/quarkus/blob/main/build-parent/pom.xml#L1162)

**Comment by Sanne** [comment]
> I agree that not hardcoding the registry everywhere seems desirable

I'm actually not sure of this being a good idea. The registry identifies which image one intends to use, and leaving it undefined opens up to various forms of supply chain attacks.

Podman was insisting in doing the right thing by refusing images whose source is unspecified; this is clearly the better behaviour unfortunately since it breaks stuff intended for Docker in which this level of precision isn't necessary this created friction in migration and the podman team ultimately gave up and changed _the default behaviour on Fedora_. 

But the option for the more secure behaviour is available still and I'd recommend to use it, and so do security guides on production systems. We should still support it.

So I'd prefer it we could keep the prefixes. Allowing an override would be welcome of course.

Regarding rate control limits, shouldn't people use proxies like they are intended to? Proxys can be quite flexible AND secured.

**Comment by gastaldi** [comment]
@Sanne has a point here. Let's keep the prefix and allow overrides in a separate PR. Thanks!

**Pull request closed without merging**